ue5osc/__init__.py

```diff
@@ -1,21 +1,123 @@
+from pathlib import Path
+from time import sleep
+
+from PIL import Image
 from pythonosc import udp_client
+from pythonosc.osc_server import BlockingOSCUDPServer
+
+from ue5osc.osc_dispatcher import OSCMessageReceiver
+
+
+class Communicator:
+    """This handles interaction between the UE5 environment and the a program."""
+
+    def __init__(self, ip: str, client_port: int, server_port: int, directory: str):
+        """Initialize OSC client and server."""
+        self.path = Path(directory)
+        self.img_number = 0
+
+        self.message_handler = OSCMessageReceiver()
+        self.client = udp_client.SimpleUDPClient(ip, client_port)
+        self.server = BlockingOSCUDPServer(
+            (ip, server_port), self.message_handler.dispatcher
+        )
+
+    def start_server(self):
+        self.server.serve_forever()
+
+    def send_and_wait(self, osc_address: str) -> object:
+        """Sends command and waits for a return value before continuing."""
+        dummy_data = 0.0
+        self.client.send_message(osc_address, dummy_data)
+        return self.message_handler.wait_for_response()
+
+    def get_project_name(self, print_val: bool = False):
+        """Returns and optionally prints the name of the current connected project."""
+        values = self.send_and_wait("/get/project")
+        if print_val:
+            self.print_values("project", values)
+        return values
 
+    def get_player_location(self, print_val: bool = False) -> list[float, float, float]:
+        """Returns x, y, z location of the player in the Unreal Environment and optionally prints."""
+        values = self.send_and_wait("/get/location")
+        if print_val:
+            self.print_values("location", values)
+        return values
 
-class OSCSender:
-    def __init__(self, ip: str, port: int):
-        self.client = udp_client.SimpleUDPClient(ip, port)
+    def set_player_location(self, x: float, y: float, z: float):
+        """Sets X, Y, and Z values of an Unreal Camera."""
+        self.client.send_message("/set/location", [x, y, z])
+
+    def get_player_rotation(self, print_val: bool = False) -> list[float, float, float]:
+        """Returns pitch, yaw, and roll and can optionally print."""
+        values = self.send_and_wait("/get/rotation")
+        if print_val:
+            self.print_values("rotation", values)
+        return values
+
+    def set_player_yaw(self, yaw: float, pawn: int = 0):
+        """Set the camera yaw in degrees."""
+        ue_roll, ue_pitch, _ = self.get_player_rotation()
+        self.client.send_message("/set/rotation", [ue_pitch, ue_roll, yaw])
+
+    def __rotate(self, rotation: float, pawn: int = 0):
+        """Rotate player a number of degrees."""
+        ue_roll, ue_pitch, ue_yaw = self.get_player_rotation()
+        yaw = float(ue_yaw) + rotation
+        self.client.send_message("/turn/left", yaw)

```

Comment:
Is this function used anywhere?

Reply:
No, it was not. I added it while I was working on boxnav and forgot to get ride of it as it is redundant.